tests/e2e/refcount_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This test suite includes refcount test on plugin side
+
+// +build runonce
+
+package e2e
+
+import (
+	"strconv"
+	"strings"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/vmware/docker-volume-vsphere/tests/utils/dockercli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/inputparams"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/misc"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/verification"
+)
+
+type RefcountTestSuite struct {
+	config *inputparams.TestConfig
+
+	volumeName        string
+	containerNameList []string
+}
+
+func (s *RefcountTestSuite) SetUpSuite(c *C) {
+	s.config = inputparams.GetTestConfig()
+	if s.config == nil {
+		c.Skip("Unable to retrieve test config, skipping refcount tests")
+	}
+}
+func (s *RefcountTestSuite) SetUpTest(c *C) {
+	s.volumeName = inputparams.GetUniqueVolumeName("RefcountTest")
+
+	// Create a volume
+	out, err := dockercli.CreateVolume(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(err, IsNil, Commentf(out))
+}
+
+func (s *RefcountTestSuite) TearDownTest(c *C) {
+
+	// If containers are leftover due to test fail, clean here
+	cName := strings.Join(s.containerNameList, " ")
+	if cName != "" {
+		out, err := dockercli.RemoveContainer(s.config.DockerHosts[0], cName)
+		c.Assert(err, IsNil, Commentf(out))
+	}
+}
+
+var _ = Suite(&RefcountTestSuite{})
+
+// Test for refcount recovery
+// * Create a volume
+// * Attach 5 containers to it with restart flag
+// * Verify attached status after each attach
+// * Check the refcount log in log file
+// * Verify the content on volume created by containers
+// * Check the volume entry on mount file : /proc/mounts
+// * Try to delete volume. Should Fail
+// * Kill Docker (ungraceful)
+// * Check the recovery log record in the log file
+// * Cleanup the containers
+// * Delete the volume. Should succeed
+// * There should be no entry in /proc/mounts
+
+func (s *RefcountTestSuite) TestRefcountingRecovery(c *C) {
+	misc.LogTestStart(c.TestName())
+
+	// Check if volume is available in docker host
+	accessible := verification.CheckVolumeAvailability(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(accessible, Equals, true, Commentf("Volume %s is not available", s.volumeName))
+
```

Comment:
General question - Any reason why we are checking for volumes only at docker host and not at esx? Or we are assuming here that things should be fine with esx?